darwin/torch/utils.py

```diff
@@ -0,0 +1,269 @@
+                        in a given image index
+    """
+    classes = {}
+    idx_to_classes = {}
+    for i, fname in enumerate(files):
+        with open(fname) as f:
+            d = json.load(f)
+            idx_to_classes[i] = []
+            for a in d["annotations"]:
+                if anno_type not in a:
+                    continue
+                cls_name = a["name"]
+                try:
+                    classes[cls_name].add(i)
+                except KeyError:
+                    classes[cls_name] = set([i])
+                if cls_name not in idx_to_classes[i]:
+                    idx_to_classes[i].append(cls_name)
+    return classes, idx_to_classes
+
+
+def fetch_darwin_dataset(
+    dataset_name: str,
+    client: Optional[Client] = None,
+    val_percentage: Optional[float] = 0.1,
+    test_percentage: Optional[float] = 0,
+    image_status: Optional[str] = "done",
+    force_fetching: Optional[bool] = False,
+    force_resplit: Optional[bool] = False,
+    split_seed: Optional[int] = None,
+):
+    """
+    Pull locally a dataset from Darwin (if needed) and create splits for
+    train, validation, and test.
+
+    Input:
+        dataset_name: name of the dataset in Darwin
+        client: Darwin client
+        val_percentage: percentage of images used in the validation set
+        test_percentage: percentage of images used in the test set
+        image_status: only pull images with under this status
+        force_fetching: discard local dataset and pull again from Darwin
+        force_resplit: discard previous split and create a new one
+        split_seed: fix seed for random split creation
+
+    Output:
+        root: local path to the dataset
+        split_path: relative path to the selected train/val/test split
+    """
+    if client is None:
+        client = Client.default()
+
+    # Get data
+    dataset = None
+    local_datasets = {dataset.slug: dataset for dataset in client.list_local_datasets()}
+    if dataset_name in local_datasets:
+        if force_fetching:
+            # Remove the local copy of the dataset
+            dbpath = Path(client.project_dir, dataset_name)
+            try:
+                shutil.rmtree(dbpath)
+            except PermissionError:
+                print(
+                    "Could not remove dataset in {dbpath}. Permission denied. \
+                      Remove it manually or disable force_fetching."
+                )
+        else:
+            dataset = local_datasets[dataset_name]
+
+    if dataset is None:  # Could not find it locally or force_fetching is True
+        remote_datasets = [dataset.slug for dataset in client.list_remote_datasets()]
+        if dataset_name in remote_datasets:
+            dataset = client.get_remote_dataset(slug=dataset_name)
+            progress, _count = dataset.pull(image_status=image_status)
+            with tqdm(total=_count, desc=f"Downloading '{dataset_name}' dataset") as pbar:
```

Comment:
Here I had a multi-threaded implementation, maybe could be useful?
```
import multiprocessing as mp
with mp.Pool(mp.cpu_count()) as pool:
   pool.map(run, progress())
```
but to work it requires a small portion of the other PR on issue #6